vktrace/src/vktrace_layer/vktrace_lib_trace.cpp

```diff
@@ -136,6 +138,161 @@ void *strip_create_extensions(const void *pNext)
     return create_info;
 }
 
+#if defined (PLATFORM_LINUX)
+
+// When quering dirty pages from /proc/self/pagemap, there is a
+// small window between the reading of this file and when the dirty bits
+// are cleared in which some other thread might modify mapped memory.
+// We trap these memory writes by setting mapped memory pages to read-only and
+// setting the signal handler for SIGSEGV. The signal handler creates a list
+// of the addresses that caused a SIGSEGV. We keep the list in sighAddrList.
+// We used the semaphore sighAddrListSem to protect access to this list.
+
+static std::list<void *> sighAddrList;
+static vktrace_sem_id sighAddrListSem = NULL;
+
+static void segvHandler(int sig, siginfo_t *si, void *ununsed)
+{
+    size_t pageSize = getpagesize();
+
+    // Note: we use sem_wait and mprotect inside a signal handler.
+    // This is not POSIX compliant, but works on Linux.
+
+    // Add the addr that caused the segv to sighAddrList
```

Comment:
It would be safer to check if the faulting address is actually in the list of mapped regions.
Then tracing an application that dies with its own SIGSEGV problem during getMappedDirtyPagesLinux would behave better.
An out of bounds address could be dealt with by either calling a non-NULL previous signal handler returned by sigaction or by flushing buffers and calling VKTRACE_FATAL_ERROR.